Force numpy prod to use 64 bit integers on Windows in some tests

mantaionut opened the pull request:
This fixes some prod and masked.prod tests on Windows. 

np.prod uses int32 on Windows so it overflows.

On Linux it uses by default int64.

Fixes #77305 
Fixes #77320 
Fixes #77334 
Fixes #77335 
Fixes #77336 
Fixes #77337

mruberry reviewed (changes requested):
Hey @mantaionut, instead of modifying the test modify the NumPy implementation of the operation associated with the operation, defined here for `prod`:

https://github.com/pytorch/pytorch/blob/34f523b22158ca4a4a7974ec867084bab98bde83/torch/testing/_internal/common_methods_invocations.py#L15998

mantaionut committed:
Force numpy prod to use 64 bit integers on Windows

This fixes some prod and masked.prod tests on Windows.
np.prod uses int32 on Windows so it overflows.
On Linux it uses by default int64.
Fixes #77305
Fixes #77320
Fixes #77334
Fixes #77335
Fixes #77336
Fixes #77337

mantaionut commented:
@mruberry I made the changes. Are those changes like what you had in mind?

mruberry commented on torch/testing/_internal/opinfo/utils.py:
```diff
@@ -258,3 +258,14 @@ def wrapper(x: np.ndarray, *args, **kwargs):
     return wrapper
+
+def numpy_prod_with_int64(): 
+    g = reference_reduction_numpy(np.prod)
```
Add a comment describing this function

mruberry commented on torch/testing/_internal/opinfo/utils.py:
```diff
@@ -258,3 +258,14 @@ def wrapper(x: np.ndarray, *args, **kwargs):
+    g = reference_reduction_numpy(np.prod)
+
+    @wraps(g)
+    def wrapper(x: np.ndarray, *args, **kwargs):
```
Why wrap `g` here? Wouldn't `numpy_prod_with_int64` already be the wrapper?

Are there other operations with similar behavior? If so, I could see making this a factory, but that could also be left for the future

mruberry commented on torch/testing/_internal/opinfo/utils.py:
```diff
@@ -258,3 +258,14 @@ def wrapper(x: np.ndarray, *args, **kwargs):
+
+    @wraps(g)
+    def wrapper(x: np.ndarray, *args, **kwargs):
+        if x.dtype == np.int8 or x.dtype == np.int16 or x.dtype == np.int32:
```
```
_low_precision_integer_dtypes = (np.int8, np.int16, np.int32)
if x.dtype in _low_precision_integer_dtypes:
```

mruberry replied:
Does this also need to check if we're running on Windows?

mantaionut replied:
I don't think it's necessary to have the windows check because it's only failing on windows. If int32 is used by default on any other platform those tests would fail there as well.

mruberry replied:
OK -- let's clarify the comment then that we want `np.prod` to compute in int64, and it does so on Linux, but not on Windows, so we just control the type promotion behavior directly

mruberry commented on torch/testing/_internal/opinfo/utils.py:
```diff
@@ -258,3 +258,14 @@ def wrapper(x: np.ndarray, *args, **kwargs):
+    @wraps(g)
+    def wrapper(x: np.ndarray, *args, **kwargs):
+        if x.dtype == np.int8 or x.dtype == np.int16 or x.dtype == np.int32:
+            kwargs["dtype"] = np.int64
```
Is this safe? What if the dtype kwarg was set explicitly?

mruberry reviewed (commented):
Yes, this is the approach I was thinking of, @mantaionut!

I made some inline comments for your review

mantaionut commented:
@mruberry I made the requested changes. The CI is failing but it does not seem to be related to my changes. Is there anything I can do to help move forward?

mruberry commented on torch/testing/_internal/opinfo/utils.py:
```diff
@@ -258,3 +258,23 @@ def wrapper(x: np.ndarray, *args, **kwargs):
 
     return wrapper
+
+def numpy_prod_with_int64():
```
What about this? 

```
def prod_numpy(a, *args, **kwargs):
  if 'dtype' not in kwargs:
    if np.issubdtype(a.dtype, np.signedinteger):
      a = a.astype(np.int64)
    elif np.issubdtype(a.dtype, np.unsignedinteger):
      a = a.astype(np.uint64)

  fn = reference_reduction_numpy(np.prod)
  return fn(a, *args, **kwargs)
      
```

Because I think NumPy promotes signed and unsigned integer types differently, and the current version of the function doesn't handle the unsigned types.

mruberry commented:
Lint test is failing -- let's just get the tests passing and then merge this!

mantaionut committed:
Changes for code review

Add comments.
Now set dtype int64 and uint64 only if its not already set.

mantaionut commented:
@mruberry  can you please approve the CI run?

mantaionut commented:
@mruberry looks like everything was successfully. Do you need me to do anything else?

mruberry reviewed (approved):
Looks great! Thanks @mantaionut!

mruberry commented:
@pytorchbot merge

mantaionut commented:
@mruberry I rebase it and the CI was successfull.

mruberry commented:
@pytorchbot merge
